docs/proposals/20220105-start-certificate-support.md

```diff
@@ -0,0 +1,269 @@
+to function in mentioned environments.
+
+## Motivation
+
+The current implementation in the Topology Operator to communicate over a TLS channel with RabbitMQ
+does not function in certain platform configurations. Ideally, users should have the flexibility to chose
+how to configure their platform, and the Topology Operator should work in reasonable configurations.
+It is almost impossible to support every possible scenario or platform configuration, however, users
+should have the flexibility to "hint" or configure the Topology Operator, so that it functions correctly
+in their platform configuration.
+
+### Goals
+
+- Design and document a solution to address the current limitation regarding secure communication.
+- Outline possible scenarios and configurations to support.
+- The proposed solution should work, regardless of how the certificate is provisioned.
+- Provide a flexible solution that works in a variety on platform configurations.
+
+### Non-Goals/Future Work
+
+- Support for specific Service Meshes.
+- A solution that only works if Cert Manager is present.
+
+## Proposal
+
+The Messaging Topology Operator constructs the URL to connect to RabbitMQ using the following pieces of
+information:
+
+- The service name from the `RabbitmqClusterReference`, as both match. 
+- The namespace name from the `RabbitmqClusterReference` (with some allow-list).
+- The protocol (HTTP or HTTPs) and port number from the Service's Ports.
+
+The above pieces of information are combined to construct the URL, such that it results in 
+`{protocol}://{service-name}.{namespace}.svc:{port}`. For example: `https://myrabbit.default.svc:15671`.
+
+A user will configure an environment variable, in the Messaging Topology Operator Deployment, specifying
+a domain name to append to the above URI. For example, an env var like `MESSAGING_DOMAIN_NAME`, with value
+`.example.com` (or `example.com`), will be appended, so that the resulting URL to connect to RabbitMQ
+would look like `https://myrabbit.default.svc.example.com:15671`. This will allow for client validation
+when RabbitMQ presents a certificate valid for `*.example.com`.
+
+![](../images/start-cert-internal-domain.jpg "Start Certificate Support with internal cluster domain")
+
+In certain scenarios, the above solution won't be sufficient. For example, RabbitMQ can be exposed via
+TCP Ingress to applications external to Kubernetes. In this scenario, RabbitMQ is provisioned with a
+certificate to match the FQDN used to connect to the Ingress, for example `rabbitmq-example.my-domain.com`.
+In this scenario, it is not possible to determine the URL from the pieces of information described above.
+In order to satisfy this scenario, the Topology Operator needs the URL as input, per RabbitMQ instance.
+The proposed solution is to use an annotation in `RabbitmqCluster`, with the URL to connect to as value.
+For example, the annotation `messaging.rabbitmq.com/override-url` with value `"https://myrabbit.example.com"`.
```

Comment:
What about the solution where we continue to connect via `{protocol}://{service-name}.{namespace}.svc:{port}` but trust certificates with a different SAN? For example, we could choose to trust `myrabbit.example.com` despite connecting to `myrabbit.mynamespace.svc`.

Reply:
I believe the cf routing release does this, since it's possible to register a route with a certificate with a different SAN, but I think that may also take advantage of controlling the entire routing system.